dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java

```diff
@@ -814,153 +823,143 @@ private TaskDefinitionLog updateTask(User loginUser, long projectCode, long task
-                    processTaskRelationMapper.queryByProcessCode(projectCode, taskRelation.getProcessDefinitionCode());
-
-            // set upstream code list
-            updateUpstreamTask(new HashSet<>(queryUpStreamTaskCodeMap.keySet()),
-                    taskCode, projectCode, taskRelation.getProcessDefinitionCode(), loginUser);
-
-            List<ProcessTaskRelation> processTaskRelationList = Lists.newArrayList(processTaskRelations);
-            List<ProcessTaskRelation> relationList = Lists.newArrayList();
-            for (ProcessTaskRelation processTaskRelation : processTaskRelationList) {
-                if (processTaskRelation.getPostTaskCode() == taskCode) {
-                    if (queryUpStreamTaskCodeMap.containsKey(processTaskRelation.getPreTaskCode())
-                            && processTaskRelation.getPreTaskCode() != 0L) {
-                        queryUpStreamTaskCodeMap.remove(processTaskRelation.getPreTaskCode());
-                    } else {
-                        processTaskRelation.setPreTaskCode(0L);
-                        processTaskRelation.setPreTaskVersion(0);
-                        relationList.add(processTaskRelation);
-                    }
+                List<TaskDefinition> upstreamTaskDefinitionList =
+                        taskDefinitionMapper.queryByCodeList(upstreamTaskCodes);
+                queryUpstreamTaskCodeMap = upstreamTaskDefinitionList.stream()
+                        .collect(Collectors.toMap(TaskDefinition::getCode, Function.identity(), (v1, v2) -> v2));
+                // upstreamTaskCodes - queryUpStreamTaskCodeMap.keySet
+                upstreamTaskCodes.removeAll(queryUpstreamTaskCodeMap.keySet());
+                if (CollectionUtils.isNotEmpty(upstreamTaskCodes)) {
+                    String notExistTaskCodes = StringUtils.join(upstreamTaskCodes, Constants.COMMA);
+                    log.error(
+                            "Some task definitions in parameter upstreamTaskCodes do not exist, notExistTaskCodes:{}.",
+                            notExistTaskCodes);
+                    putMsg(result, Status.TASK_DEFINE_NOT_EXIST, notExistTaskCodes);
+                    return result;
                 }
             }
-            processTaskRelationList.removeAll(relationList);
-            for (Map.Entry<Long, TaskDefinition> queryUpStreamTask : queryUpStreamTaskCodeMap.entrySet()) {
-                taskRelation.setPreTaskCode(queryUpStreamTask.getKey());
-                taskRelation.setPreTaskVersion(queryUpStreamTask.getValue().getVersion());
-                processTaskRelationList.add(taskRelation);
-            }
-            if (MapUtils.isEmpty(queryUpStreamTaskCodeMap) && CollectionUtils.isNotEmpty(processTaskRelationList)) {
-                processTaskRelationList.add(processTaskRelationList.get(0));
-            }
         }
+
+        updateUpstreamTaskAndDag(queryUpstreamTaskCodeMap, taskDefinitionToUpdate, projectCode, loginUser);
```

Comment:
Can we simply add one line to switch version here?
```java
switchVersion(loginUser, projectCode, taskCode, taskDefinitionToUpdate.getVersion());
```
Maybe this fix is more simple, and can also decouple the logic of update task and update dag.